Team Matchmaker Integration

BlackYps opened the pull request:
Don't really know what to write here. You know what it is...

Geosearchef committed:
team matchmaking WIP

Geosearchef committed:
team matchmaking WIP

Geosearchef committed:
team matchmaking WIP

Geosearchef committed:
team matchmaking WIP

Geosearchef committed:
team matchmaking WIP

Geosearchef committed:
team matchmaking WIP

Geosearchef committed:
team matchmaking WIP

Geosearchef committed:
team matchmaking WIP

Geosearchef committed:
team matchmaking WIP

Geosearchef committed:
team matchmaking WIP

Geosearchef committed:
party ready system

Geosearchef committed:
party structure rework, factions WIP

Geosearchef committed:
party factions UI, UI restructure

Geosearchef committed:
party system WIP

Geosearchef committed:
party system FACTIONS WIP

Geosearchef committed:
party system factions

Geosearchef committed:
do not show faction buttons on disable

1-alex98 committed:
Clean up and Exception fixes

1-alex98 committed:
WIP

Geosearchef committed:
add context menu to party

Geosearchef committed:
player card ui

Geosearchef committed:
restructure TMM ui

Geosearchef committed:
TMM queue list display

Geosearchef committed:
tmm queue UI, timer

Geosearchef committed:
tmm queue - business logic

Geosearchef committed:
ensure game path is valid before sending invite

BlackYps committed:
Integrate new matchmaker ui

BlackYps committed:
Use i18n instead of hardcoded text

BlackYps committed:
Filter player out of party

BlackYps committed:
Refactor country flag loading

BlackYps committed:
Change ready button to play button

Geosearchef committed:
hide ready button, disable join queue buttons based on party size and ownership, hide invite player button

Geosearchef committed:
fix tests

Geosearchef committed:
TESTS DISABLED, REVERT!

Geosearchef committed:
WIP remove faction from matchmaker queue join, faction order fix, temporary ugly GameLaunchMessage parsing fix, temporary matchmaker game launch

Geosearchef committed:
switch faction buttons, fix flag on faction select, force faction selection

Geosearchef committed:
tmm - display matching status

Geosearchef committed:
i18n queue name

Geosearchef committed:
matchmaking - use faction ids instead of boolean array

Geosearchef committed:
notify and block party invite / join when in queue

Geosearchef committed:
sanity checks, game launch cleanup

BlackYps committed:
Uppercase now controlled by styleclass to enable override by themes

BlackYps committed:
Delete obsolete play button

BlackYps committed:
Enable resizing of the UI

BlackYps committed:
Add new heading for queue area

BlackYps committed:
Fix some bugs

BlackYps committed:
Restyle queue cards

BlackYps committed:
Implement queue name lookup and shortening

BlackYps committed:
Replace rating labels with league icon + label

micheljung committed:
Add VCS config

BlackYps committed:
Layout changes and playercard grows properly

BlackYps committed:
Use proper string formatting

BlackYps committed:
Add matching status labels

BlackYps committed:
Merge branch 'feature/team-matchmaking' into feature/new-tmm-ui

# Conflicts:
#	src/main/java/com/faforever/client/teammatchmaking/MatchmakingQueueItemController.java
#	src/main/java/com/faforever/client/teammatchmaking/PartyMemberItemController.java
#	src/main/resources/i18n/messages.properties
#	src/main/resources/theme/play/team_matchmaking.fxml
#	src/main/resources/theme/play/teammatchmaking/matchmaking_member_card.fxml
#	src/main/resources/theme/play/teammatchmaking/matchmaking_queue_card.fxml

BlackYps committed:
Fix merge errors

Geosearchef committed:
tmm: parse factions as string, fix status indicators for own faction selection

Geosearchef committed:
fix queue join race condition, set queue joined button state based on server state

Geosearchef committed:
fix faction parsing for game launch message

Geosearchef committed:
fix install4j github release url

Geosearchef committed:
remove / disable ladder1v1 tests

Geosearchef committed:
bump install4j version WIP

Geosearchef committed:
unbump install4j version (to 7, use different key)

Geosearchef committed:
send own factions to server upon entering queue

Geosearchef committed:
select all factions in an empty party

BlackYps committed:
Fix empty slot in party member pane

BlackYps committed:
Highlight party owner

BlackYps committed:
Align kick button to the right

BlackYps committed:
Change queues heading text when player is not the party owner

BlackYps committed:
Show invited players in invite dialog

BlackYps committed:
Add scrollpane to matchmaker ui

Geosearchef committed:
fix disabled queue on startup

Geosearchef committed:
catch game cancel during launch when multiqueueing

BlackYps committed:
Add leader highlight for own player

Geosearchef committed:
fix already in queue when starting a custom game

Geosearchef committed:
uncheck queue button when launch failed due to being in game

BlackYps committed:
Move splitPane divider to 0.7

Geosearchef committed:
Merge branch 'develop' into feature/team-matchmaking

# Conflicts:
#	ci/github-release.sh
#	src/main/java/com/faforever/client/game/GameService.java
#	src/main/java/com/faforever/client/play/PlayController.java
#	src/main/java/com/faforever/client/remote/domain/GameInfoMessage.java
#	src/main/resources/font/dfc-icons.ttf
#	src/main/resources/i18n/messages.properties
#	src/main/resources/theme/style.css
#	src/test/java/com/faforever/client/game/GameInfoMessageBuilder.java
#	src/test/java/com/faforever/client/rankedmatch/Ladder1V1ControllerTest.java

Geosearchef committed:
tmm merge fixes

BlackYps committed:
Adapt icons to new workflow

BlackYps committed:
Load division icon instead of avatar

BlackYps committed:
Add chat functionality WIP

BlackYps committed:
Spawn new chat tabs only when necessary

BlackYps committed:
Add chat functionality

BlackYps committed:
Make code cleaner by removing checks for uppercase style class

BlackYps committed:
Code cleanup (mostly deleting of redundant code)

BlackYps committed:
Remove ladder tab

BlackYps committed:
Remove logic for ladder searches

BlackYps committed:
Remove notification for match available

BlackYps committed:
Fix bug in PartyMemberItemController

BlackYps committed:
Disable chat tab highlight on party message

BlackYps committed:
Merge branch 'develop' into feature/team-matchmaking

# Conflicts:
#	src/main/java/com/faforever/client/game/GameService.java
#	src/main/java/com/faforever/client/rankedmatch/Ladder1v1Controller.java
#	src/main/resources/theme/settings/settings.fxml
#	src/test/java/com/faforever/client/game/GameServiceTest.java

BlackYps committed:
Make chat position dynamic

BlackYps committed:
Show team matchmaker map pools (#2032)

BlackYps committed:
UI adjustments

BlackYps committed:
Fix failing tests

BlackYps committed:
Only highlight tab on new messages

BlackYps committed:
Fix error in log on queue deselect

BlackYps committed:
Add tests for TeamMatchmakingService

BlackYps committed:
Add indication when partyMember is in a game

BlackYps committed:
Merge branch 'develop' into feature/team-matchmaking

# Conflicts:
#	src/main/java/com/faforever/client/api/FafApiAccessorImpl.java
#	src/main/java/com/faforever/client/main/MainController.java
#	src/main/resources/theme/play/ranked_1v1.fxml

BlackYps committed:
Fix merge errors

Sheikah45 commented:
Can you clean up the commit messages?

Brutus5000 commented on src/main/java/com/faforever/client/api/FafApiAccessorImpl.java:
```diff
@@ -431,12 +432,24 @@ public void deleteModVersionReview(String id) {
-            "mapVersion.map.versions.reviews.player"));
-    return new Tuple<>(jsonApiDoc.get(), jsonApiDoc.getMeta());
+  public List<MatchmakerQueueMapPool> getMatchmakerPools(int matchmakerQueueId) {
+    return getAll("/data/matchmakerQueueMapPool", ImmutableMap.of(
```
just Map.of()

BlackYps replied:
All other functions in that class also use ImmutableMap, should those be changed too?

Brutus5000 replied:
Yes. Map.of is also immutable and is java standard, so we should use this.

Brutus5000 replied:
It came after Java 8 into the JDK that's why you find guava ImmutableMap/List/... everywhere in old code.

Brutus5000 commented on src/main/java/com/faforever/client/api/FafApiAccessorImpl.java:
```diff
@@ -431,12 +432,24 @@ public void deleteModVersionReview(String id) {
+
+  @Override
+  public Optional<MatchmakerQueue> getMatchmakerQueue(String technicalName) {
+    List<MatchmakerQueue> queue = getAll("/data/matchmakerQueue", ImmutableMap.of(
```
just Map.of()

Brutus5000 commented on src/main/java/com/faforever/client/api/MockFafApiAccessor.java:
```diff
@@ -284,8 +285,13 @@ public void deleteModVersionReview(String id) {
-  public Tuple<List<Ladder1v1Map>, java.util.Map<String, ?>> getLadder1v1MapsWithMeta(int count, int page) {
-    return new Tuple(Collections.emptyList(), Collections.emptyMap());
+  public List<MatchmakerQueueMapPool> getMatchmakerPools(int matchmakerQueueId) {
+    return Collections.emptyList();
```
just List.of()

Brutus5000 commented on src/main/java/com/faforever/client/api/dto/Leaderboard.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.time.OffsetDateTime;
+
+@Getter
+@Setter
+@EqualsAndHashCode(of = "id")
```
Why not `@Data` or `@Value`?

BlackYps replied:
I think I copied that from another dto class. No real reason to not use `@Data`

Brutus5000 commented on src/main/java/com/faforever/client/api/dto/MapPool.java:
```diff
@@ -7,13 +7,19 @@
+
 @Getter
 @Setter
 @EqualsAndHashCode(of = "id")
```
Why not @Data or @Value?

Brutus5000 commented on src/main/java/com/faforever/client/chat/ChatController.java:
```diff
@@ -178,6 +178,8 @@ private void onConnectionStateChange(ConnectionState newValue) {
     Platform.runLater(() -> {
       ChatMessage message = event.getMessage();
+      if (message.getSource() != null && message.getSource().endsWith("'sParty"))
+        return;
```
.endsWith("'sParty") has an implicit semantical meaning and therefore should be put into a dedicated method.

I'd guess something like `boolean isOwnPartiesChannel(String name)`
But just comparing the end of the name sounds really... odd?

Brutus5000 commented on src/main/java/com/faforever/client/chat/ChatChannelUser.java:
```diff
@@ -45,7 +45,7 @@
   }
 
-  ChatChannelUser(String username, Color color, boolean moderator, Player player) {
+  public ChatChannelUser(String username, Color color, boolean moderator, Player player) {
```
Why is this public now? Could be a code smell. Not sure.

BlackYps replied:
This and related changes were done in a214cbdf by Geosearchef. This is where he implemented the hacky way to display the contextmenu. I can't say for sure, but this looks like a candidate for refactoring to make the whole thing cleaner

Brutus5000 commented on src/main/java/com/faforever/client/api/dto/MatchmakerQueueMapPool.java:
```diff
@@ -0,0 +1,28 @@
+
+@Getter
+@Setter
+@EqualsAndHashCode(of = "id")
```
same here

Brutus5000 commented on src/main/java/com/faforever/client/api/dto/MatchmakerQueue.java:
```diff
@@ -0,0 +1,28 @@
+
+@Getter
+@Setter
+@EqualsAndHashCode(of = "id")
```
same here

Brutus5000 commented on src/main/java/com/faforever/client/chat/ChatUserContextMenuController.java:
```diff
@@ -145,7 +145,7 @@ public void initialize() {
   }
 
-  ContextMenu getContextMenu() {
+  public ContextMenu getContextMenu() {
```
Why does this method even exist if there is `getRoot()`? I know it was there before, but maybe you can explain.

BlackYps replied:
No idea...

BlackYps replied:
This was for the hacky implementation in the partyMemberItemController

Brutus5000 commented on src/main/java/com/faforever/client/chat/MatchmakingChatController.java:
```diff
@@ -0,0 +1,116 @@
+  private Channel channel;
+  private MapChangeListener<String, ChatChannelUser> usersChangeListener;
+
+  @Inject
```
Remove. `@Inject` isn't needed.

Brutus5000 commented on src/main/java/com/faforever/client/chat/MatchmakingChatController.java:
```diff
@@ -0,0 +1,116 @@
+  private MapChangeListener<String, ChatChannelUser> usersChangeListener;
+
+  @Inject
+  // TODO cut dependencies
```
Is there a TODO or not? If yes, please fix, if no please remove.

BlackYps replied:
I don't know. The other two classes that extend from AbstractChatTabController also have it and I wanted to stay consistent.

Sheikah45 replied:
I think the thing to do here would be to cut the unnecessary dependencies.

BlackYps replied:
This is a separate issue. This todo also exists in the other chatcontrollers. I put it here for consistency

Brutus5000 commented on src/main/java/com/faforever/client/game/GameService.java:
```diff
@@ -331,9 +339,17 @@ public ReadOnlyBooleanProperty gameRunningProperty() {
+          Collections.singletonList(new Action(i18n.get("dismiss")))
+      ));
+      return completedFuture(null);
+    }
```
duplicate code, please put into dedicated method

Brutus5000 commented on src/main/java/com/faforever/client/game/GameService.java:
```diff
@@ -534,27 +546,27 @@ public void addOnRankedMatchNotificationListener(Consumer<MatchmakerInfoMessage>
+  public void onMatchmakerSearchStopped() {
+    if (inMatchmakerQueue.get()) {
+      fafService.stopSearchMatchmaker();
+      inMatchmakerQueue.set(false);
     }
   }
```
I suggest debug logging of both cases

BlackYps replied:
The other case is that inMatchmakerQueue is already false?

Brutus5000 commented on src/main/java/com/faforever/client/rankedmatch/MatchmakerInfoMessage.java:
```diff
@@ -9,32 +9,38 @@
 
+  private String action; // TODO: doesn't exist anymore
+
   public static class MatchmakerQueue {
```
Use Lombok `@Data` instead?
Get rid of all the getter and setter below.

Brutus5000 commented on src/main/java/com/faforever/client/remote/FafService.java:
```diff
@@ -96,16 +96,17 @@
 
-  public void stopSearchingRanked() {
-    fafServerAccessor.stopSearchingRanked();
+  //TODO: use for TMM or remove
```
Please resolve the todo

Brutus5000 commented on src/main/java/com/faforever/client/remote/FafService.java:
```diff
@@ -458,13 +459,27 @@ public void restoreGameSession(int id) {
-        .parallelStream()
-        .map(ladder1v1Map -> MapBean.fromMapVersionDto(ladder1v1Map.getMapVersion()))
+  public CompletableFuture<Tuple<List<MapBean>, Integer>> getMatchmakerMapsWithPageCount(int matchmakerQueueId, int count, int page) {
+    List<MapVersion> mapVersions = fafApiAccessor.getMatchmakerPools(matchmakerQueueId)
+        .stream()
+        .map(MatchmakerQueueMapPool::getMapPool)
+        .flatMap(mapPool -> mapPool.getMapVersions().stream())
+        .distinct()
+        .collect(toList());
+    int totalPages = (mapVersions.size() - 1) / count + 1;
+    return CompletableFuture.completedFuture(new Tuple<>(mapVersions
+        .stream()
+        .skip((page - 1) * count)
+        .limit(count)
+        .map(MapBean::fromMapVersionDto)
         .collect(toList()),
-        ((HashMap<String,Integer>) tuple.getSecond().get("page")).get("totalPages")));
+        totalPages));
+  }
+
```
I know it was there before, but now it looks even more like a code smell :thinking:

BlackYps replied:
What exactly? Or the whole thing?

Brutus5000 replied:
The integer that is returned as well. Calculating page sizes / page numbers is not a problem tied to map beans, it applies to every collection and it could go into some util function.

Sheikah45 replied:
Just as a note for all other api calls the api gives us the total page number. However it doesnt do that for the mapPool maps currently which is why it needs to be calculated here.

BlackYps replied:
I agree, but all the other api calls use pagination to limit the api call itself. This one is the only one (yet) that has to limit the returning result. What would an appropriate place for the helper function be?

Brutus5000 commented on src/main/java/com/faforever/client/remote/domain/AcceptPartyInviteMessage.java:
```diff
@@ -0,0 +1,19 @@
+  public void setSender_id(Integer sender_id) {
+    this.sender_id = sender_id;
+  }
+}
```
Underscores in variables and methods names are a no-go.

Brutus5000 commented on src/main/java/com/faforever/client/remote/domain/GameMatchmakingMessage.java:
```diff
@@ -0,0 +1,17 @@
+    this.state = state;
+  }
+
+}
```
Underscores in variables are a no-go.

Brutus5000 commented on src/main/java/com/faforever/client/remote/domain/InviteToPartyMessage.java:
```diff
@@ -0,0 +1,19 @@
+  public void setRecipient_id(Integer recipient_id) {
+    this.recipient_id = recipient_id;
+  }
+}
```
stop it

Brutus5000 commented on src/main/java/com/faforever/client/remote/domain/KickPlayerFromPartyMessage.java:
```diff
@@ -0,0 +1,19 @@
+  public void setKicked_player_id(Integer kicked_player_id) {
+    this.kicked_player_id = kicked_player_id;
+  }
+}
```
my eyes!! :eyes:

Brutus5000 commented on src/main/java/com/faforever/client/remote/domain/MatchmakingState.java:
```diff
@@ -0,0 +1,21 @@
+  }
+
+  public static MatchmakingState fromString(String string) {
+    return Arrays.stream(MatchmakingState.values()).filter(s -> s.getString().equals(string)).findFirst().orElse(null);
```
Please put line breaks after each stream operation

Brutus5000 commented on src/main/java/com/faforever/client/remote/domain/PartyInviteMessage.java:
```diff
@@ -0,0 +1,18 @@
+
+  public void setSender(Integer sender) {
+    this.sender = sender;
+  }
```
Use lombok instead

Brutus5000 commented on src/main/java/com/faforever/client/remote/domain/SetPartyFactionsMessage.java:
```diff
@@ -0,0 +1,24 @@
+
+  public void setFactions(List<String> factions) {
+    this.factions = factions;
+  }
```
Use lombok instead

Brutus5000 commented on src/main/java/com/faforever/client/teammatchmaking/Party.java:
```diff
@@ -0,0 +1,96 @@
+    playerService.getPlayersByIds(Collections.singletonList(message.getOwner()))
+        .thenAccept(players -> {
+          if (!players.isEmpty()) {
+            Platform.runLater(() -> owner.set(players.get(0)));
```
Wait what? An object isn't allowed to change itself outside of the main thread?

Sheikah45 replied:
This seems fishy it should be able to run outside the main thread unless it is bound to some graphic somewhere

BlackYps replied:
There are a lot of graphical bindings to this, so it is necessary

Brutus5000 commented on src/main/java/com/faforever/client/teammatchmaking/Party.java:
```diff
@@ -0,0 +1,96 @@
+  }
+
+  public void fromInfoMessage(PartyInfoMessage message, PlayerService playerService) {
+    playerService.getPlayersByIds(Collections.singletonList(message.getOwner()))
```
use `List.of(message.getOwner())`

Brutus5000 commented on src/main/java/com/faforever/client/teammatchmaking/Party.java:
```diff
@@ -0,0 +1,96 @@
+    members = FXCollections.observableArrayList();
+  }
+
+  public void fromInfoMessage(PartyInfoMessage message, PlayerService playerService) {
```
This is a heavy violation of separation of concerns. This method is NOT allowed in a dto. Put it into the playerservice or wherever you need it.

Brutus5000 commented on src/main/java/com/faforever/client/teammatchmaking/Party.java:
```diff
@@ -0,0 +1,96 @@
+          }).filter(Objects::nonNull).collect(Collectors.toList());
+          //TODO: this is a race condition. The api might answer with big delay and then server message order might be changed.
+          Platform.runLater(() -> this.members.setAll(members));
+        });
```
The nested complexitiy of these lines is too damn high. Outsource it in functions or at least improve formatting with proper linebreaking on each stream.

Brutus5000 commented on src/test/java/com/faforever/client/chat/AbstractChatTabControllerTest.java:
```diff
@@ -233,6 +234,7 @@ public void testOnChatMessageAction() {
   }
 
   @Test
+  @Ignore // TODO: REVERT
```
Revert revert! :D

Brutus5000 commented on src/test/java/com/faforever/client/game/GameServiceTest.java:
```diff
@@ -484,80 +479,83 @@ public void testOnGameInfoRemove() {
-
-    assertThat(instance.searching1v1Property().get(), is(false));
-  }
+  //TODO: remove/reimplement TMM
```
I guess this block can be removed now.

BlackYps replied:
same here

Brutus5000 commented on src/test/java/com/faforever/client/game/GameServiceTest.java:
```diff
@@ -629,12 +627,13 @@ public void runWithLiveReplayIfNoGameSet() {
-    instance.startSearchLadder1v1(null);
-    verify(eventBus).post(any(GameDirectoryChooseEvent.class));
-  }
+  //TODO: remove/reimplement TMM
```
Can be removed now as well?

Sheikah45 replied:
I think it needs to be reimplemented otherwise there is no test for this I think

BlackYps replied:
Reimplementing this was surprisingly easy

Brutus5000 commented on src/test/java/com/faforever/client/teammatchmaking/TeamMatchmakingServiceTest.java:
```diff
@@ -0,0 +1,240 @@
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+public class TeamMatchmakingServiceTest extends AbstractPlainJavaFxTest {
```
I see a lot of empty tests there.

Brutus5000 commented on src/main/java/com/faforever/client/teammatchmaking/InvitePlayerController.java:
```diff
@@ -0,0 +1,102 @@
+            return false;
+          }
+
+          if (playerTextField.getText().trim().equals("")) {
```
`playerTextField.getText().isBlank()`

Brutus5000 commented on src/main/java/com/faforever/client/teammatchmaking/InvitePlayerController.java:
```diff
@@ -0,0 +1,102 @@
+    });
+
+
+    //TODO: use longest common subsequence instead and sort list
```
Is this todo still relevant?

BlackYps replied:
seems to be already implemented

Sheikah45 replied:
So the TODO should be removed then

Brutus5000 commented on src/main/java/com/faforever/client/teammatchmaking/InvitePlayerController.java:
```diff
@@ -0,0 +1,102 @@
+    });
+
+
+    //TODO: use longest common subsequence instead and sort list
+    filteredPlayerList.predicateProperty().bind(Bindings.createObjectBinding(() -> p -> {
+          if (playerService.getCurrentPlayer().map(Player::getUsername).map(n -> n.equals(p)).orElse(true)) {
+            return false;
+          }
+
+          if (playerTextField.getText().trim().equals("")) {
+            return playerService.getPlayerForUsername(p).map(player -> player.getSocialStatus() == SocialStatus.FRIEND).orElse(false);
+          } else {
+            return p.toLowerCase().contains(playerTextField.getText().toLowerCase());
+          }
+        }, playerTextField.textProperty()
```
Line breaks in the stream operations improves readability

Brutus5000 commented on src/main/java/com/faforever/client/teammatchmaking/InvitePlayerController.java:
```diff
@@ -0,0 +1,102 @@
+        }, playerTextField.textProperty()
+    ));
+
+    //TODO
```
Remove?

Brutus5000 commented on src/main/java/com/faforever/client/teammatchmaking/MatchmakingQueueItemController.java:
```diff
@@ -0,0 +1,151 @@
+    this.queue = queue;
+
+    playersInQueueLabel.textProperty().bind(createStringBinding(
+        () -> i18n.get("teammatchmaking.playersInQueue", queue.getPlayersInQueue()).toUpperCase(),
```
Isn't the toUpperCase something that should be configurable in the control via css or something?

Brutus5000 commented on src/main/java/com/faforever/client/teammatchmaking/PartyMemberItemController.java:
```diff
@@ -0,0 +1,176 @@
+  public ImageView playerStatusImageView;
+
+  private Player player;
+  //TODO: this is a bit hacky
```
Don't tell me it's hacky without telling what and why. I assume the WeakReference is meant here, but it I don't know for sure.

BlackYps replied:
I don't know either. That comment is from Geosearchef.

BlackYps replied:
a214cbdf

Sheikah45 replied:
The contextmenu is separable from the chatUser which might be the thing to do here as I don't understand why the chatUser is necessary if all we want is a context menu

Sheikah45 replied:
This also seems like it would be a different context menu

Brutus5000 commented on src/main/java/com/faforever/client/teammatchmaking/PartyMemberItemController.java:
```diff
@@ -0,0 +1,176 @@
+    chatUser = new ChatChannelUser(player.getUsername(), chatService.getChatUserColor(player.getUsername()), false, player);
+
+    countryImageView.imageProperty().bind(createObjectBinding(() -> countryFlagService.loadCountryFlag(
+        StringUtils.isEmpty(player.getCountry()) ? "" : player.getCountry()).orElse(null), player.countryProperty()));
```
`StringUtils.isEmpty(player.getCountry()) ? "" : player.getCountry()`

If the string is empty it's empty? Thanks captain.

BlackYps replied:
This seems to be a nullcheck, no?

Brutus5000 replied:
Yes, but the CountryFlagService.loadCountryFlag has null handling builtin ^^

BlackYps commented:
squashing the WIP commits is a major act, because we have some merge commits from develop in this branch. In theory you can use `rebase --interactive --rebase-berges` to preserve the merge commits, so you don't have to resolve the conflicts again, but it doesn't seem to be working. I still got conflicts to resolve, so I left it, because this seriously goes beyond my git abilities.

Brutus5000 commented on src/main/java/com/faforever/client/teammatchmaking/PartyMemberItemController.java:
```diff
@@ -0,0 +1,176 @@
+
+  private ChatChannelUser chatUser;
+
+  public PartyMemberItemController(CountryFlagService countryFlagService, AvatarService avatarService, PlayerService playerService, TeamMatchmakingService teamMatchmakingService, UiService uiService, ChatService chatService, I18n i18n) {
```
Use `@RequiredArgsConstructor`

Brutus5000 commented on src/main/java/com/faforever/client/teammatchmaking/TeamMatchmakingController.java:
```diff
@@ -0,0 +1,287 @@
+    player = playerService.getCurrentPlayer().get();
+    initializeUppercaseText();
+    countryImageView.imageProperty().bind(createObjectBinding(() -> countryFlagService.loadCountryFlag(
+        StringUtils.isEmpty(player.getCountry()) ? "" : player.getCountry()).orElse(null), player.countryProperty()));
```
Again here `StringUtils.isEmpty(player.getCountry()) ? "" : player.getCountry()` seems  useless.
And duplicate code

Brutus5000 commented on src/main/java/com/faforever/client/teammatchmaking/TeamMatchmakingController.java:
```diff
@@ -0,0 +1,287 @@
+  }
+
+  private void initializeUppercaseText() {
+    matchmakerHeadingLabel.setText(i18n.get("teammatchmaking.playerTitle").toUpperCase());
+    partyHeadingLabel.setText(i18n.get("teammatchmaking.partyTitle").toUpperCase());
+    invitePlayerButton.setText(i18n.get("teammatchmaking.invitePlayer").toUpperCase());
+    leavePartyButton.setText(i18n.get("teammatchmaking.leaveParty").toUpperCase());
```
Can this not be done via css? Then it would be completely redundant (handled in FXML)

BlackYps replied:
I wish! CSS can do that, but the "special" css for javafx cannot :(

Brutus5000 replied:
Shame

Brutus5000 commented on src/main/java/com/faforever/client/teammatchmaking/TeamMatchmakingController.java:
```diff
@@ -0,0 +1,287 @@
+  public void onChatMessage(ChatMessageEvent event) {
+    Platform.runLater(() -> {
+      ChatMessage message = event.getMessage();
+      if (message.getSource().equals(String.format("#%s'sParty", teamMatchmakingService.getParty().getOwner().getUsername()))) {
```
As stated earlier, this should be outsourced into a dedicated function

Brutus5000 commented on src/main/java/com/faforever/client/teammatchmaking/TeamMatchmakingService.java:
```diff
@@ -0,0 +1,363 @@
+  private boolean queuesAdded = false;
+
+  @Override
+  public void afterPropertiesSet() throws Exception {
```
If not needed remove Initializing bean or use `@RequiredArgsConstructor` and do all the current constructor logic here.

Brutus5000 commented on src/main/java/com/faforever/client/teammatchmaking/TeamMatchmakingService.java:
```diff
@@ -0,0 +1,363 @@
+    List<CompletableFuture<?>> futures = new ArrayList<>();
+    message.getQueues().forEach(remoteQueue -> {
+      MatchmakingQueue localQueue = matchmakingQueues.stream()
+          .filter(q -> Objects.equals(q.getQueueName(), remoteQueue.getQueueName())).findFirst().orElse(null);
```
line breaks after the stream operators

Brutus5000 commented on src/main/java/com/faforever/client/teammatchmaking/TeamMatchmakingService.java:
```diff
@@ -0,0 +1,363 @@
+      return;
+    }
+
+    matchmakingQueues.stream().filter(q -> q.getMatchingStatus() != null).forEach(q -> {
```
linebreaks after each stream operator

Brutus5000 reviewed (changes requested):
There is a quite a high amount of TODOs that should be either resolved or removed.

BlackYps committed:
Refactoring

Use Map instead of ImmutableMap
Use List.of instead of Collections.emptyList
Clean imports
Extract method in GameService
Use line breaks in stream operations
Controller refactoring

BlackYps committed:
Improve invite dialog

Sheikah45 commented on src/main/java/com/faforever/client/game/GameService.java:
```diff
@@ -505,27 +518,23 @@ public Game getByUid(int uid) {
+    inMatchmakerQueue.set(true);
 
-    return modService.getFeaturedMod(LADDER_1V1.getTechnicalName())
+    return modService.getFeaturedMod(FAF.getTechnicalName())//TODO: use matchmaking featured mod
```
Can this TODO be resolved?

BlackYps replied:
FAF is the right featured mod for the matchmaker. I will delete the todo

Sheikah45 commented on src/main/java/com/faforever/client/game/GameService.java:
```diff
@@ -534,27 +543,27 @@ public void addOnRankedMatchNotificationListener(Consumer<MatchmakerInfoMessage>
               gameLaunchMessage.getArgs().add("/startspot " + gameLaunchMessage.getMapPosition());
 
-              startGame(gameLaunchMessage, faction, RatingMode.LADDER_1V1);
+              startGame(gameLaunchMessage, gameLaunchMessage.getFaction(), NONE); // TODO: rating mode
```
This one seems needed as well unless I am missing something

BlackYps replied:
This is for displaying the right rating in the game. We should grab the appropriate leaderboard from the queue information once #1791 is implemented.

Sheikah45 commented on src/main/java/com/faforever/client/rankedmatch/MatchmakerInfoMessage.java:
```diff
@@ -4,86 +4,47 @@
   }
-
-  private String action;
+  @Getter
+  @Setter
```
I believe these can be removed with the @Data at top

BlackYps replied:
The @Data annotation is only for the nested class

Sheikah45 commented on src/main/java/com/faforever/client/teammatchmaking/MatchmakingQueueItemController.java:
```diff
@@ -0,0 +1,151 @@
+        () -> i18n.get("teammatchmaking.playersInQueue", queue.getPlayersInQueue()).toUpperCase(),
+        queue.playersInQueueProperty()));
+
+    matchFoundLabel.visibleProperty().bind(matchFoundLabel.managedProperty());
+    matchStartingLabel.visibleProperty().bind(matchStartingLabel.managedProperty());
+    matchCancelledLabel.visibleProperty().bind(matchCancelledLabel.managedProperty());
```
JavaFxUtil.bindManagedToVisible

Sheikah45 commented on src/main/java/com/faforever/client/teammatchmaking/PartyMemberItemController.java:
```diff
@@ -0,0 +1,167 @@
+          playerStatusImageView.setVisible(true);
+          playerCard.pseudoClassStateChanged(PLAYING_PSEUDO_CLASS, true);
+        });
+      }
```
We need to make sure this listener is removed if a player leaves the party

BlackYps replied:
Everytime a partymember leaves or joines the player cards get rerendered, that means it clears all children from the pane and instantiates new controllers. I am not sure, if this also cleans up the controllers, or if we need to do some form of garbage collection

Sheikah45 reviewed (changes requested):
First pass. There are quite a few tests that deal directly with playing the game and launch states that should be reimplemented for tmm still.

BlackYps committed:
Cache api call and fix queue display bugs

BlackYps committed:
Add logging when matchmaker search stopped

BlackYps committed:
Fix missing info on partymember cards

BlackYps committed:
Refactoring

Move method from Party.java to TeamMatchmakingService
Refactor partychat logic
Cleanup todos
Reposition kick button

BlackYps commented:
I resolved many of the comments. For the rest of them I could use some help, because they seem to be more complex.

BlackYps committed:
Add github actions

1-alex98 commented on .github/workflows/checks.yml:
```diff
@@ -0,0 +1,40 @@
+# This workflow will build a Java project with Gradle
+# For more information see: https://help.github.com/actions/language-and-framework-guides/building-and-testing-java-with-gradle
+
+name: Checks
+
+on:
+  push:
+    branches: [ develop ]
+  pull_request:
+    branches: [ develop ]
+
+jobs:
+  checks:
+
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Set up JDK 15.0.1
+        uses: actions/setup-java@v1
+        with:
+          java-version: 15.0.1
+          java-package: jdk+fx
+
```
Should not be part of this PR

Sheikah45 replied:
as a note this was just added to allow for releasing via github actions for the test today and should disappear once the branch is rebased on develop

1-alex98 commented on src/main/java/com/faforever/client/remote/MockFafServerAccessor.java:
```diff
@@ -245,17 +234,17 @@ public void requestMatchmakerInfo() {
     GameLaunchMessage gameLaunchMessage = new GameLaunchMessage();
     gameLaunchMessage.setUid(123);
-    gameLaunchMessage.setMod(KnownFeaturedMod.DEFAULT.getTechnicalName());
+    gameLaunchMessage.setMod(KnownFeaturedMod.DEFAULT.getTechnicalName()); // TODO
```
TODO what also refer to an issue if you make a TODO

BlackYps replied:
I am not sure, I guess it refers to changing the gamemode to a matchmaking mode. However we use FAF gamemode, so I guess this todo is obsolete.

1-alex98 commented on src/main/java/com/faforever/client/teammatchmaking/MatchmakingQueueItemController.java:
```diff
@@ -0,0 +1,144 @@
+    mapPoolButton.setText(i18n.get("teammatchmaking.mapPool").toUpperCase());
+
+    Timeline queuePopTimeUpdater = new Timeline(1, new KeyFrame(javafx.util.Duration.seconds(0), (ActionEvent event) -> {
+      if (queue.getQueuePopTime() != null) {
+        Instant now = Instant.now();
+        Duration timeUntilPopQueue = Duration.between(now, queue.getQueuePopTime());
+        if (!timeUntilPopQueue.isNegative()) {
+          queuePopTimeLabel.setText(i18n.get("teammatchmaking.queuePopTimer",
+              timeUntilPopQueue.toMinutes(),
+              timeUntilPopQueue.toSecondsPart()).toUpperCase());
+        }
+      }
```
Extract into method

1-alex98 commented on src/main/java/com/faforever/client/teammatchmaking/MatchmakingQueueItemController.java:
```diff
@@ -0,0 +1,144 @@
+    return queueItemRoot;
+  }
+
+  public void setQueue(MatchmakingQueue queue) {
```
the method is a little long

1-alex98 commented on src/main/java/com/faforever/client/teammatchmaking/TeamMatchmakingController.java:
```diff
@@ -0,0 +1,296 @@
+    initializeBindings();
+
+    teamMatchmakingService.getParty().getMembers().addListener((Observable o) -> {
+      playerCard.pseudoClassStateChanged(LEADER_PSEUDO_CLASS,
+          (teamMatchmakingService.getParty().getOwner().equals(player) && teamMatchmakingService.getParty().getMembers().size() > 1));
+      List<PartyMember> members = new ArrayList<>(teamMatchmakingService.getParty().getMembers());
+      partyMemberPane.getChildren().clear();
+      members.removeIf(partyMember -> partyMember.getPlayer().equals(player));
+      for(int i = 0; i < members.size(); i++) {
+        PartyMemberItemController controller = uiService.loadFxml("theme/play/teammatchmaking/matchmaking_member_card.fxml");
+        controller.setMember(members.get(i));
+        if (members.size() == 1) {
+          partyMemberPane.add(controller.getRoot(), 0, 0, 2, 1);
+        } else {
+          partyMemberPane.add(controller.getRoot(), i % 2, i / 2);
+        }
+      }
```
extract method

BlackYps committed:
Disable queue buttons when someone is in a game

BlackYps committed:
Change queue heading message when you are in queue

BlackYps committed:
Refactoring

BlackYps committed:
Create new context menu for party members

BlackYps committed:
Add more tests

BlackYps committed:
Remember factions

BlackYps committed:
merge branch 'develop' into feature/team-matchmaking

BlackYps committed:
code cleanup

Sheikah45 commented on src/test/java/com/faforever/client/teammatchmaking/TeamMatchmakingServiceTest.java:
```diff
@@ -0,0 +1,410 @@
+  }
+
+  @Test
+  public void testOnPartyInfoMessage() {
+
+  }
```
empty test

Sheikah45 reviewed (commented):
I think this is in a good place from my perspective. There was still one empty test

BlackYps committed:
Add invite item to chat context menu

BlackYps committed:
Add tests

BlackYps committed:
Fix queues not displayed bug

1-alex98 commented on src/main/java/com/faforever/client/chat/ChatController.java:
```diff
@@ -135,10 +136,10 @@ public void initialize() {
 
     chatService.addChannelsListener(change -> {
-      if (change.wasRemoved()) {
+      if (change.wasRemoved() && !change.getValueRemoved().getName().endsWith("'sParty")) {
```
Needs to be a constant

Brutus5000 replied:
Overload isMatchmakerPartyMessage to take a String as well. Then use the dedicated function

Brutus5000 commented on src/main/java/com/faforever/client/api/FafApiAccessorImpl.java:
```diff
@@ -427,28 +427,41 @@ public void deleteModVersionReview(String id) {
-            "mapVersion.map.statistics,mapVersion.map.reviewsSummary,mapVersion.map.versions.reviews," +
-            "mapVersion.map.versions.reviews.player"));
-    return new Tuple<>(jsonApiDoc.get(), jsonApiDoc.getMeta());
+  public List<MatchmakerQueueMapPool> getMatchmakerPools(int matchmakerQueueId) {
```
Should this be cached as well?

BlackYps replied:
probably a good idea

Brutus5000 commented on src/main/java/com/faforever/client/api/dto/Leaderboard.java:
```diff
@@ -0,0 +1,18 @@
+  private String id;
+  private OffsetDateTime createTime;
+  private OffsetDateTime updateTime;
+  private String description_key;
+  private String name_key;
+  private String technical_name;
```
We just changed this in the API to be properly camel case again

Sheikah45 replied:
Brutus did this get changed to being camelCase after you reverted?

BlackYps committed:
Code cleanup

BlackYps committed:
Cache map pools

BlackYps committed:
Merge branch 'develop' into feature/team-matchmaking

# Conflicts:
#	src/main/java/com/faforever/client/api/FafApiAccessorImpl.java

BlackYps committed:
leaderboard to camelCase

Sheikah45 commented on src/main/java/com/faforever/client/api/FafApiAccessorImpl.java:
```diff
@@ -446,28 +446,42 @@ public void deleteModVersionReview(String id) {
+  @Cacheable(value = CacheNames.MATCHMAKER_QUEUES, sync = true)
+  public Optional<MatchmakerQueue> getMatchmakerQueue(String technicalName) {
+    List<MatchmakerQueue> queue = getAll("/data/matchmakerQueue", java.util.Map.of(
+        FILTER, rsql(qBuilder().string("technicalName").eq(technicalName))));
```
I would include the leaderboard of the queue, although if the api did not change back to camelcase this may break it at the moment

Sheikah45 commented on src/main/java/com/faforever/client/teammatchmaking/TeamMatchmakingController.java:
```diff
@@ -0,0 +1,324 @@
+      if (matchmakingChatController != null) {
+        matchmakingChatController.closeChannel();
+      }
+      createChannelTab(String.format("#%s'sParty", newValue.getUsername()));
```
This should be a constant as well ideally related to the one used for checking the channel name

Sheikah45 commented on src/main/java/com/faforever/client/teammatchmaking/TeamMatchmakingController.java:
```diff
@@ -0,0 +1,324 @@
+      }
+      createChannelTab(String.format("#%s'sParty", newValue.getUsername()));
+    });
+    createChannelTab(String.format("#%s'sParty", teamMatchmakingService.getParty().getOwner().getUsername()));
```
constant

Sheikah45 commented on src/main/java/com/faforever/client/chat/ChatController.java:
```diff
@@ -186,6 +189,14 @@ public void onChatMessage(ChatMessageEvent event) {
+  }
+
+  private boolean isMatchmakerPartyMessage(String channelName) {
+    return channelName.endsWith("'sParty");
```
Would be good to have this as a constant to make referencing easier

BlackYps committed:
Use constant for party channels

BlackYps committed:
Include leaderboard in queue api call

1-alex98 commented on src/main/java/com/faforever/client/teammatchmaking/TeamMatchmakingController.java:
```diff
@@ -0,0 +1,325 @@
+    });
+  }
+
+  private synchronized void renderQueues() {
```
why would that be synchronized? I mean it just executes in site PlatformRun.later ... That make is synchronized because there is only one UI Thread !!!

BlackYps replied:
That is leftover from my attempts to fix a bug with concurrent access to the queue list. Using the synchronized list should fix it. I'll delete it

1-alex98 commented on src/main/java/com/faforever/client/teammatchmaking/TeamMatchmakingController.java:
```diff
@@ -0,0 +1,325 @@
+    Platform.runLater(() -> {
+      ChatMessage message = event.getMessage();
+      if (message.getSource().equals(matchmakingChatController.getReceiver())) {
+        matchmakingChatController.onChatMessage(message);
```
Since there might be a lot of messages received here. Only execute this very line on the UI  Thread no need to switch to the UI thread to check if the message is releavnt => 99% less execution on the UI thread for this action

1-alex98 commented on src/main/java/com/faforever/client/teammatchmaking/TeamMatchmakingController.java:
```diff
@@ -0,0 +1,325 @@
+  }
+
+  public void onFactionButtonClicked(ActionEvent actionEvent) {
+
```
remove new line

1-alex98 commented on src/main/java/com/faforever/client/teammatchmaking/PartyMemberItemController.java:
```diff
@@ -0,0 +1,166 @@
+    usernameLabel.textProperty().bind(player.usernameProperty());
+    leagueLabel.textProperty().bind(createStringBinding(
+        () -> i18n.get("leaderboard.divisionName", RatingUtil.getLeaderboardRating(player)).toUpperCase(),
+        player.globalRatingMeanProperty())); // TODO: replace this with divisionproperty once it is available
```
Create an issue and reference it

BlackYps replied:
The leagueservice has already been merged into faf stack. I think it is just a matter of time for deployment now. Or maybe there needs to be extra work done to create the api endpoints. I am not sure.

BlackYps committed:
Code cleanup

BlackYps committed:
Delete leftover message
